### PR TITLE
MINOR Improve logging during the ZK to KRaft migration

### PR DESCRIPTION
**Pull request by mumrah**
Several small improvements to logging during the migration.

* Add an exponential backoff to from 100ms to 1min while the controller is waiting for brokers to show up
* Increases one-time logs to INFO
* Adds a summary of the migration records
* Use RecordRedactor for summary of migration batches (TRACE only)

### CODE REVIEW

File: metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java [hunk 1]
```diff
@@ -158,7 +185,7 @@ private void recoverMigrationStateFromZK() {
     private boolean isControllerQuorumReadyForMigration() {
         Optional<String> notReadyMsg = this.quorumFeatures.reasonAllControllersZkMigrationNotReady();
         if (notReadyMsg.isPresent()) {
-            log.info("Still waiting for all controller nodes ready to begin the migration. due to:" + notReadyMsg.get());
+            log.info("Still waiting for all controller nodes ready to begin the migration. Not ready due to:" + notReadyMsg.get());
```
Comment (cmccabe) [inline metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java]: Should be `WARN`?

File: metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java [hunk 2]
```diff
@@ -690,36 +727,18 @@ public void run() throws Exception {
...
-                sb.append(")");
-                sb.append(" at version ");
-                sb.append(apiMessageAndVersion.version());
-                sb.append(")");
-                return sb.toString();
-            } else {
-                return apiMessageAndVersion.toString();
-            }
-        }).collect(Collectors.joining(","));
+    String batchToRedactedString(Collection<ApiMessageAndVersion> batch) {
```
Comment (cmccabe) [inline metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java]: Hmm, I think one record per log line would be better. Too hard to follow otherwise. And we might end up creating some super-giant strings that would break stuff.

**Comment by mumrah** [comment]
From an integration test 

summary:
```
Completed migration of metadata from ZooKeeper to KRaft. 71 records were generated in 180 ms across 4 batches. The record types were {PARTITION_RECORD=2, CONFIG_RECORD=2, TOPIC_RECORD=2, PRODUCER_IDS_RECORD=1}. The current metadata offset is now 91 with an epoch of 1. Saw 3 brokers in the migrated metadata [0, 1, 2]
```

all logs during migration (TRACE enabled)

```
[2023-07-13 16:56:18,029] INFO [KRaftMigrationDriver id=3000] Starting ZK migration (org.apache.kafka.metadata.migration.KRaftMigrationDriver:606)
[2023-07-13 16:56:18,036] INFO [KRaftMigrationDriver id=3000] Migrating 5 records from ZK (org.apache.kafka.metadata.migration.KRaftMigrationDriver:612)
[2023-07-13 16:56:18,037] TRACE [KRaftMigrationDriver id=3000] TopicRecord(name='test', topicId=K16VvIw3TLmaYnQ6XjK4PQ) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,037] TRACE [KRaftMigrationDriver id=3000] ConfigRecord(resourceType=2, resourceName='test', name='segment.bytes', value='102400') (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,037] TRACE [KRaftMigrationDriver id=3000] ConfigRecord(resourceType=2, resourceName='test', name='segment.ms', value='300000') (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,037] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=0, topicId=K16VvIw3TLmaYnQ6XjK4PQ, replicas=[0, 1, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,037] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=1, topicId=K16VvIw3TLmaYnQ6XjK4PQ, replicas=[2, 0, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,038] INFO [KRaftMigrationDriver id=3000] Waiting for the metadata layer to commit migration record batch (org.apache.kafka.metadata.migration.KRaftMigrationDriver:52)
[2023-07-13 16:56:18,072] INFO [KRaftMigrationDriver id=3000] Finished waiting for the metadata layer to commit migration record batch (org.apache.kafka.metadata.migration.KRaftMigrationDriver:55)
[2023-07-13 16:56:18,074] INFO [KRaftMigrationDriver id=3000] Migrating 9 records from ZK (org.apache.kafka.metadata.migration.KRaftMigrationDriver:612)
[2023-07-13 16:56:18,074] TRACE [KRaftMigrationDriver id=3000] TopicRecord(name='__consumer_offsets', topicId=P42iU37kTlaOybvxSQXRiA) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,074] TRACE [KRaftMigrationDriver id=3000] ConfigRecord(resourceType=2, resourceName='__consumer_offsets', name='compression.type', value='producer') (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,074] TRACE [KRaftMigrationDriver id=3000] ConfigRecord(resourceType=2, resourceName='__consumer_offsets', name='cleanup.policy', value='compact') (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,074] TRACE [KRaftMigrationDriver id=3000] ConfigRecord(resourceType=2, resourceName='__consumer_offsets', name='segment.bytes', value='104857600') (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,074] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=4, topicId=P42iU37kTlaOybvxSQXRiA, replicas=[2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=2, partitionEpoch=2) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,074] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=3, topicId=P42iU37kTlaOybvxSQXRiA, replicas=[1], isr=[1], removingReplicas=[], addingReplicas=[], leader=1, leaderRecoveryState=0, leaderEpoch=3, partitionEpoch=3) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,074] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=2, topicId=P42iU37kTlaOybvxSQXRiA, replicas=[0], isr=[0], removingReplicas=[], addingReplicas=[], leader=0, leaderRecoveryState=0, leaderEpoch=3, partitionEpoch=3) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,074] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=0, topicId=P42iU37kTlaOybvxSQXRiA, replicas=[1], isr=[1], removingReplicas=[], addingReplicas=[], leader=1, leaderRecoveryState=0, leaderEpoch=3, partitionEpoch=3) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,074] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=1, topicId=P42iU37kTlaOybvxSQXRiA, replicas=[2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=2, partitionEpoch=2) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,074] INFO [KRaftMigrationDriver id=3000] Waiting for the metadata layer to commit migration record batch (org.apache.kafka.metadata.migration.KRaftMigrationDriver:52)
[2023-07-13 16:56:18,101] INFO [KRaftMigrationDriver id=3000] Finished waiting for the metadata layer to commit migration record batch (org.apache.kafka.metadata.migration.KRaftMigrationDriver:55)
[2023-07-13 16:56:18,105] INFO [KRaftMigrationDriver id=3000] Migrating 56 records from ZK (org.apache.kafka.metadata.migration.KRaftMigrationDriver:612)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] TopicRecord(name='__transaction_state', topicId=xF_T1xqNR9ifNRODyU2PxA) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] ConfigRecord(resourceType=2, resourceName='__transaction_state', name='compression.type', value='uncompressed') (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] ConfigRecord(resourceType=2, resourceName='__transaction_state', name='cleanup.policy', value='compact') (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] ConfigRecord(resourceType=2, resourceName='__transaction_state', name='unclean.leader.election.enable', value='false') (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] ConfigRecord(resourceType=2, resourceName='__transaction_state', name='min.insync.replicas', value='2') (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] ConfigRecord(resourceType=2, resourceName='__transaction_state', name='segment.bytes', value='104857600') (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=42, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 1, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=7, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 0, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=13, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 0, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=0, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 1, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=37, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 0, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=6, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 1, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,105] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=32, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 2, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=18, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 1, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=40, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 1, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=31, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 0, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=45, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 2, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=15, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 2, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=12, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 1, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=46, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 1, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=48, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 1, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=49, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 0, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=28, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 1, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=2, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 2, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=20, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 2, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=24, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 1, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=3, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 2, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=21, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 2, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=29, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 0, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=39, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 2, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=38, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 2, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=14, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 2, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=10, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 1, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=44, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 2, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=9, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 2, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=22, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 1, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=43, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 0, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=4, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 1, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=30, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 1, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=33, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 2, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=25, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 0, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=17, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 0, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=23, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 0, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=47, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 0, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=26, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 2, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=36, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 1, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=5, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 0, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=8, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 2, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=16, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 1, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=11, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 0, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=19, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 0, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=27, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[0, 2, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=41, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 0, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=1, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 0, 1], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=34, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[2, 1, 0], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] TRACE [KRaftMigrationDriver id=3000] PartitionRecord(partitionId=35, topicId=xF_T1xqNR9ifNRODyU2PxA, replicas=[1, 0, 2], isr=[2], removingReplicas=[], addingReplicas=[], leader=2, leaderRecoveryState=0, leaderEpoch=4, partitionEpoch=4) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,106] INFO [KRaftMigrationDriver id=3000] Waiting for the metadata layer to commit migration record batch (org.apache.kafka.metadata.migration.KRaftMigrationDriver:52)
[2023-07-13 16:56:18,136] INFO [KRaftMigrationDriver id=3000] Finished waiting for the metadata layer to commit migration record batch (org.apache.kafka.metadata.migration.KRaftMigrationDriver:55)
[2023-07-13 16:56:18,139] INFO [KRaftMigrationDriver id=3000] Migrating 1 records from ZK (org.apache.kafka.metadata.migration.KRaftMigrationDriver:612)
[2023-07-13 16:56:18,139] TRACE [KRaftMigrationDriver id=3000] ProducerIdsRecord(brokerId=0, brokerEpoch=-1, nextProducerId=0) (org.apache.kafka.metadata.migration.KRaftMigrationDriver:615)
[2023-07-13 16:56:18,139] INFO [KRaftMigrationDriver id=3000] Waiting for the metadata layer to commit migration record batch (org.apache.kafka.metadata.migration.KRaftMigrationDriver:52)
[2023-07-13 16:56:18,171] INFO [KRaftMigrationDriver id=3000] Finished waiting for the metadata layer to commit migration record batch (org.apache.kafka.metadata.migration.KRaftMigrationDriver:55)
[2023-07-13 16:56:18,182] INFO [KRaftMigrationDriver id=3000] Waiting for the metadata layer to complete the migration (org.apache.kafka.metadata.migration.KRaftMigrationDriver:52)
[2023-07-13 16:56:18,211] INFO [KRaftMigrationDriver id=3000] Finished waiting for the metadata layer to complete the migration (org.apache.kafka.metadata.migration.KRaftMigrationDriver:55)
[2023-07-13 16:56:18,211] INFO [KRaftMigrationDriver id=3000] Completed migration of metadata from ZooKeeper to KRaft. 71 records were generated in 180 ms across 4 batches. The record types were {PARTITION_RECORD=2, CONFIG_RECORD=2, TOPIC_RECORD=2, PRODUCER_IDS_RECORD=1}. The current metadata offset is now 91 with an epoch of 1. Saw 3 brokers in the migrated metadata [0, 1, 2]. (org.apache.kafka.metadata.migration.KRaftMigrationDriver:633)
[2023-07-13 16:56:18,212] INFO [KRaftMigrationDriver id=3000] Finished initial migration of ZK metadata to KRaft. Transitioned migration state from ZkMigrationLeadershipState{kraftControllerId=3000, kraftControllerEpoch=1, kraftMetadataOffset=-1, kraftMetadataEpoch=-1, lastUpdatedTimeMs=1689281772391, migrationZkVersion=0, controllerZkEpoch=5, controllerZkVersion=5} to ZkMigrationLeadershipState{kraftControllerId=3000, kraftControllerEpoch=1, kraftMetadataOffset=91, kraftMetadataEpoch=1, lastUpdatedTimeMs=1689281772391, migrationZkVersion=1, controllerZkEpoch=5, controllerZkVersion=5} (org.apache.kafka.metadata.migration.KRaftMigrationDriver:250)
```

**Review by cmccabe — APPROVED** [review]

**Comment by mumrah** [comment]
Test failures seem to be unrelated flaky

image
